File: Python version/skin_analysis/analysis.py
from __future__ import annotations

from dataclasses import replace

import numpy as np
import pandas as pd

from .config import (
    BASELINE_RISE_MIN_RUN,
    BASELINE_RISE_ROLLING_POINTS,
    BASELINE_TAIL_POINTS,
    DATA_COL,
    DEFAULT_BASELINE_DURATION_SEC,
    DEFAULT_BASELINE_WARNING_THRESHOLD_PCT,
    DEFAULT_DRUG_APPLY_TIME_SEC,
    DEFAULT_DRUG_APPLY_TOLERANCE_SEC,
    DROP_SIGMA_THRESHOLD,
    DT_SEC,
    LEGACY_DROP_SEARCH_POINTS,
    PBS_BASELINE_PRE_ROLL_POINTS,
)
from .models import BaselineWarningStatus, CurveSegment, CurveSplit, DropDetectionSource, ProcessedSignal


def read_xlsx_single(path: str) -> pd.DataFrame | None:
    try:
        df = pd.read_excel(path, engine="openpyxl")
    except Exception:
        return None

    if DATA_COL not in df.columns:
        return None
    return df[[DATA_COL]].copy()


def _offset_pct(value: float, baseline: float) -> float:
    if abs(baseline) < 1e-12:
        return 0.0
    return ((value / baseline) * 100.0) - 100.0


def _seconds_to_points(duration_sec: float, minimum: int = 0) -> int:
    return max(minimum, int(round(duration_sec / DT_SEC)))


def _first_threshold_crossing(
    samples: np.ndarray,
    threshold: float,
    start_idx: int,
    end_idx: int,
) -> int | None:
    if start_idx >= end_idx:
        return None

    search_range = samples[start_idx:end_idx]
    over_idx = np.where(search_range > threshold)[0]
    if over_idx.size == 0:
        return None
    return start_idx + int(over_idx[0])


def _max_continuous_rise_offset_pct(baseline_segment: np.ndarray, baseline_avg: float) -> float:
    if baseline_segment.size < BASELINE_RISE_ROLLING_POINTS:
        return 0.0

    kernel = np.ones(BASELINE_RISE_ROLLING_POINTS, dtype=float) / BASELINE_RISE_ROLLING_POINTS
    rolling_mean = np.convolve(baseline_segment, kernel, mode="valid")
    if rolling_mean.size < BASELINE_RISE_MIN_RUN:
        return 0.0

    run_length = 1
    max_offset_pct = 0.0

    for index in range(1, len(rolling_mean)):
        if rolling_mean[index] > rolling_mean[index - 1]:
            run_length += 1
            continue

        if run_length >= BASELINE_RISE_MIN_RUN:
            max_offset_pct = max(max_offset_pct, _offset_pct(float(rolling_mean[index - 1]), baseline_avg))
        run_length = 1

    if run_length >= BASELINE_RISE_MIN_RUN:
        max_offset_pct = max(max_offset_pct, _offset_pct(float(rolling_mean[-1]), baseline_avg))

    return max(0.0, max_offset_pct)


def _baseline_warning_status(tail_hit: bool, rise_hit: bool) -> BaselineWarningStatus:
    hit_count = int(tail_hit) + int(rise_hit)
    if hit_count >= 2:
        return "inaccurate"
    if hit_count == 1:
        return "warning"
    return "ok"


def _effective_baseline_points(
    sample_count: int,
    requested_baseline_duration_sec: float,
    apply_window_start_idx: int,
) -> tuple[int, bool]:
    requested_baseline_points = min(
        sample_count,
        _seconds_to_points(max(0.0, requested_baseline_duration_sec), minimum=1),
    )
    if requested_baseline_points <= apply_window_start_idx:
        return requested_baseline_points, False

    shortened_points = min(sample_count, apply_window_start_idx)
    if shortened_points <= 0:
        return min(sample_count, 1), True
    return shortened_points, True


def _build_timing_warning_details(
    requested_baseline_duration_sec: float,
    effective_baseline_duration_sec: float,
    baseline_was_auto_shortened: bool,
    apply_time_sec: float,
    apply_tolerance_sec: float,
    drop_search_fallback_used: bool,
) -> tuple[str, ...]:
    details: list[str] = []

    if baseline_was_auto_shortened:
        if effective_baseline_duration_sec <= DT_SEC:
            details.append(
                "Baseline overlapped the apply window and was clamped to the minimum baseline length."
            )
        else:
            details.append(
                f"Baseline shortened from {requested_baseline_duration_sec:.1f}s to "
                f"{effective_baseline_duration_sec:.1f}s to stay before the apply window."
            )

    if drop_search_fallback_used:
        details.append(
            f"No threshold crossing found within {apply_time_sec:.1f}s +/- {apply_tolerance_sec:.1f}s; "
            "used automatic fallback search."
        )

    return tuple(details)


def _fallback_drop_index(samples: np.ndarray, threshold: float, baseline_end_idx: int) -> int:
    search_end = min(len(samples), baseline_end_idx + LEGACY_DROP_SEARCH_POINTS)
    crossing_idx = _first_threshold_crossing(samples, threshold, baseline_end_idx, search_end)
    if crossing_idx is None:
        return baseline_end_idx
    return crossing_idx


def split_index_from_time_sec(split_time_sec: float, sample_count: int) -> int:
    if sample_count < 2:
        return 0
    requested_index = int(round(max(0.0, float(split_time_sec)) / DT_SEC))
    return max(1, min(sample_count - 1, requested_index))


def _split_index_for_data(curve_split: CurveSplit, sample_count: int) -> int:
    if sample_count < 2:
        return 0
    return max(1, min(sample_count - 1, int(curve_split.split_index)))


def pbs_pre_roll_window_indices(curve_split: CurveSplit, sample_count: int) -> tuple[int, int, int]:
    split_index = _split_index_for_data(curve_split, sample_count)
    if split_index <= 0:
        return 0, 0, 0
    start_index = max(0, split_index - PBS_BASELINE_PRE_ROLL_POINTS)
    return start_index, split_index, split_index - start_index


def _pbs_pre_roll_warning(pre_roll_points: int) -> str | None:
    if pre_roll_points >= PBS_BASELINE_PRE_ROLL_POINTS:
        return None
    return f"PBS baseline pre-roll {pre_roll_points}/{PBS_BASELINE_PRE_ROLL_POINTS} points before split."


def _build_static_segment_signal(
    samples: np.ndarray,
    baseline_warning_threshold_pct: float,
    baseline_duration_sec: float,
) -> ProcessedSignal | None:
    samples = np.asarray(samples, dtype=float)
    if samples.size == 0:
        return None

    warning_threshold_pct = max(0.0, float(baseline_warning_threshold_pct))
    requested_baseline_duration_sec = max(float(baseline_duration_sec), DT_SEC)
    baseline_len = min(
        len(samples),
        _seconds_to_points(requested_baseline_duration_sec, minimum=1),
    )
    baseline_segment = samples[:baseline_len]
    initial_avg = float(np.mean(baseline_segment))
    effective_baseline_duration_sec = baseline_len * DT_SEC

    tail_len = min(len(baseline_segment), BASELINE_TAIL_POINTS)
    tail_mean = float(np.mean(baseline_segment[-tail_len:])) if tail_len > 0 else initial_avg
    tail_offset_pct = _offset_pct(tail_mean, initial_avg)
    rise_offset_pct = _max_continuous_rise_offset_pct(baseline_segment, initial_avg)
    tail_warning_hit = abs(tail_offset_pct) >= warning_threshold_pct
    rise_warning_hit = rise_offset_pct >= warning_threshold_pct
    warning_status = _baseline_warning_status(tail_warning_hit, rise_warning_hit)
    final_avg = float(np.nanmean(samples[-100:])) if len(samples) >= 100 else float("nan")
    delta_raw = final_avg - initial_avg if not np.isnan(final_avg) else float("nan")

    return ProcessedSignal(
        time_sec=np.arange(len(samples), dtype=float) * DT_SEC,
        capacitance=samples,
        drop_time=0.0,
        delta_capacitance=delta_raw,
        initial_avg=initial_avg,
        effective_baseline_points=baseline_len,
        effective_baseline_duration_sec=effective_baseline_duration_sec,
        baseline_was_auto_shortened=False,
        drop_detection_source="window",
        drop_search_fallback_used=False,
        timing_warning_details=(),
        baseline_warning_status=warning_status,
        baseline_tail_offset_pct=tail_offset_pct,
        baseline_rise_offset_pct=rise_offset_pct,
        baseline_tail_warning_hit=tail_warning_hit,
        baseline_rise_warning_hit=rise_warning_hit,
    )


def _samples_for_segment(
    samples: np.ndarray,
    curve_split: CurveSplit | None,
    segment: CurveSegment,
) -> tuple[np.ndarray | None, int]:
    if segment == "full" or curve_split is None:
        if segment == "lanolin":
            return None, 0
        return samples, 0

    split_index = _split_index_for_data(curve_split, len(samples))
    if split_index <= 0:
        return None, 0
    if segment == "lanolin":
        return samples[:split_index], 0
    if segment == "pbs":
        start_index, _split_index, pre_roll_points = pbs_pre_roll_window_indices(curve_split, len(samples))
        return samples[start_index:], pre_roll_points
    return samples, 0


def analyze_signal(
    data: np.ndarray,
    baseline_warning_threshold_pct: float = DEFAULT_BASELINE_WARNING_THRESHOLD_PCT,
    baseline_duration_sec: float = DEFAULT_BASELINE_DURATION_SEC,
    drug_apply_time_sec: float = DEFAULT_DRUG_APPLY_TIME_SEC,
    drug_apply_tolerance_sec: float = DEFAULT_DRUG_APPLY_TOLERANCE_SEC,
) -> ProcessedSignal | None:
    samples = np.asarray(data, dtype=float)
    if samples.size == 0:
        return None

    warning_threshold_pct = max(0.0, float(baseline_warning_threshold_pct))
    requested_baseline_duration_sec = max(float(baseline_duration_sec), DT_SEC)
    requested_apply_time_sec = max(0.0, float(drug_apply_time_sec))
    requested_apply_tolerance_sec = max(0.0, float(drug_apply_tolerance_sec))

    apply_center_idx = _seconds_to_points(requested_apply_time_sec)
    apply_tolerance_points = _seconds_to_points(requested_apply_tolerance_sec)
    apply_window_start_idx = max(0, apply_center_idx - apply_tolerance_points)
    apply_window_end_idx = min(len(samples), apply_center_idx + apply_tolerance_points + 1)

    baseline_len, baseline_was_auto_shortened = _effective_baseline_points(
        len(samples),
        requested_baseline_duration_sec,
        apply_window_start_idx,
    )
    baseline_segment = samples[:baseline_len]
    initial_avg = float(np.mean(baseline_segment))
    sample_std = float(np.std(baseline_segment, ddof=1)) if baseline_len > 1 else 0.0
    effective_baseline_duration_sec = baseline_len * DT_SEC

    tail_len = min(len(baseline_segment), BASELINE_TAIL_POINTS)
    tail_mean = float(np.mean(baseline_segment[-tail_len:])) if tail_len > 0 else initial_avg
    tail_offset_pct = _offset_pct(tail_mean, initial_avg)
    rise_offset_pct = _max_continuous_rise_offset_pct(baseline_segment, initial_avg)
    tail_warning_hit = abs(tail_offset_pct) >= warning_threshold_pct
    rise_warning_hit = rise_offset_pct >= warning_threshold_pct
    warning_status = _baseline_warning_status(tail_warning_hit, rise_warning_hit)

    threshold = initial_avg + DROP_SIGMA_THRESHOLD * max(sample_std, 1e-4)
    window_search_start_idx = max(baseline_len, apply_window_start_idx)
    idx_drop = _first_threshold_crossing(
        samples,
        threshold,
        window_search_start_idx,
        apply_window_end_idx,
    )
    drop_detection_source: DropDetectionSource = "window"
    drop_search_fallback_used = idx_drop is None

    if idx_drop is None:
        idx_drop = _fallback_drop_index(samples, threshold, baseline_len)
        drop_detection_source = "fallback_auto"

    final_avg = float(np.nanmean(samples[-100:])) if len(samples) >= 100 else float("nan")
    delta_raw = final_avg - initial_avg if not np.isnan(final_avg) else float("nan")
    timing_warning_details = _build_timing_warning_details(
        requested_baseline_duration_sec=requested_baseline_duration_sec,
        effective_baseline_duration_sec=effective_baseline_duration_sec,
        baseline_was_auto_shortened=baseline_was_auto_shortened,
        apply_time_sec=requested_apply_time_sec,
        apply_tolerance_sec=requested_apply_tolerance_sec,
        drop_search_fallback_used=drop_search_fallback_used,
    )

    return ProcessedSignal(
        time_sec=np.arange(len(samples), dtype=float) * DT_SEC,
        capacitance=samples,
        drop_time=idx_drop * DT_SEC,
        delta_capacitance=delta_raw,
        initial_avg=initial_avg,
        effective_baseline_points=baseline_len,
        effective_baseline_duration_sec=effective_baseline_duration_sec,
        baseline_was_auto_shortened=baseline_was_auto_shortened,
        drop_detection_source=drop_detection_source,
        drop_search_fallback_used=drop_search_fallback_used,
        timing_warning_details=timing_warning_details,
        baseline_warning_status=warning_status,
        baseline_tail_offset_pct=tail_offset_pct,
        baseline_rise_offset_pct=rise_offset_pct,
        baseline_tail_warning_hit=tail_warning_hit,
        baseline_rise_warning_hit=rise_warning_hit,
    )


def process_single_file(
    file_path: str,
    baseline_warning_threshold_pct: float = DEFAULT_BASELINE_WARNING_THRESHOLD_PCT,
    baseline_duration_sec: float = DEFAULT_BASELINE_DURATION_SEC,
    drug_apply_time_sec: float = DEFAULT_DRUG_APPLY_TIME_SEC,
    drug_apply_tolerance_sec: float = DEFAULT_DRUG_APPLY_TOLERANCE_SEC,
    curve_split: CurveSplit | None = None,
    segment: CurveSegment = "pbs",
) -> ProcessedSignal | None:
    df = read_xlsx_single(file_path)
    if df is None or df.empty:
        return None
    samples = df[DATA_COL].to_numpy()
    segment_samples, pre_roll_points = _samples_for_segment(samples, curve_split, segment)
    if segment_samples is None or len(segment_samples) == 0:
        return None
    if segment == "lanolin":
        return _build_static_segment_signal(
            segment_samples,
            baseline_warning_threshold_pct=baseline_warning_threshold_pct,
            baseline_duration_sec=baseline_duration_sec,
        )
    adjusted_baseline_duration_sec = baseline_duration_sec
    adjusted_drug_apply_time_sec = drug_apply_time_sec
    pre_roll_warning = None
    if segment == "pbs" and curve_split is not None:
        adjusted_baseline_duration_sec = PBS_BASELINE_PRE_ROLL_POINTS * DT_SEC
        adjusted_drug_apply_time_sec = drug_apply_time_sec + (pre_roll_points * DT_SEC)
        pre_roll_warning = _pbs_pre_roll_warning(pre_roll_points)

    signal = analyze_signal(
        segment_samples,
        baseline_warning_threshold_pct=baseline_warning_threshold_pct,
        baseline_duration_sec=adjusted_baseline_duration_sec,
        drug_apply_time_sec=adjusted_drug_apply_time_sec,
        drug_apply_tolerance_sec=drug_apply_tolerance_sec,
    )
    if signal is None or pre_roll_warning is None:
        return signal
    return replace(
        signal,
        timing_warning_details=(pre_roll_warning, *signal.timing_warning_details),
    )

File: desktop-tauri/src-tauri/src/analysis.rs
use std::path::Path;

use calamine::{open_workbook_auto, DataType, Reader};

use crate::config::{
    BASELINE_RISE_MIN_RUN, BASELINE_RISE_ROLLING_POINTS, BASELINE_TAIL_POINTS, DATA_COL,
    DROP_SIGMA_THRESHOLD, DT_SEC, LEGACY_DROP_SEARCH_POINTS,
};
use crate::models::{AnalysisParams, ProcessedSignal};

// ── Numeric helpers ────────────────────────────────────────────────────────────

fn mean(values: &[f64]) -> f64 {
    if values.is_empty() {
        return f64::NAN;
    }
    values.iter().sum::<f64>() / values.len() as f64
}

fn sample_std(values: &[f64]) -> f64 {
    if values.len() < 2 {
        return 0.0;
    }
    let avg = mean(values);
    let variance = values.iter().map(|v| (v - avg).powi(2)).sum::<f64>() / (values.len() - 1) as f64;
    variance.sqrt()
}

fn nanmean(values: &[f64]) -> f64 {
    let (sum, count) = values.iter().filter(|v| !v.is_nan()).fold((0.0, 0usize), |(s, c), &v| (s + v, c + 1));
    if count == 0 { f64::NAN } else { sum / count as f64 }
}

fn offset_pct(value: f64, baseline: f64) -> f64 {
    if baseline.abs() < 1e-12 {
        0.0
    } else {
        (value / baseline) * 100.0 - 100.0
    }
}

fn seconds_to_points(sec: f64) -> usize {
    (sec / DT_SEC).round() as usize
}

fn cell_to_f64(cell: &impl DataType) -> f64 {
    if let Some(v) = cell.get_float() {
        v
    } else if let Some(v) = cell.get_int() {
        v as f64
    } else if let Some(s) = cell.get_string() {
        s.parse::<f64>().unwrap_or(f64::NAN)
    } else if let Some(b) = cell.get_bool() {
        if b { 1.0 } else { 0.0 }
    } else {
        f64::NAN
    }
}

// ── Baseline accuracy ──────────────────────────────────────────────────────────

fn max_continuous_rise_offset_pct(baseline_segment: &[f64], baseline_avg: f64) -> f64 {
    let n = baseline_segment.len();
    if n < BASELINE_RISE_ROLLING_POINTS {
        return 0.0;
    }
    let rolling_len = n - BASELINE_RISE_ROLLING_POINTS + 1;
    let rolling_means: Vec<f64> = (0..rolling_len)
        .map(|i| mean(&baseline_segment[i..i + BASELINE_RISE_ROLLING_POINTS]))
        .collect();
    if rolling_means.len() < BASELINE_RISE_MIN_RUN {
        return 0.0;
    }
    let mut run_length = 1usize;
    let mut max_offset = 0.0f64;
    for i in 1..rolling_means.len() {
        if rolling_means[i] > rolling_means[i - 1] {
            run_length += 1;
        } else {
            if run_length >= BASELINE_RISE_MIN_RUN {
                max_offset = max_offset.max(offset_pct(rolling_means[i - 1], baseline_avg));
            }
            run_length = 1;
        }
    }
    if run_length >= BASELINE_RISE_MIN_RUN {
        if let Some(&last) = rolling_means.last() {
            max_offset = max_offset.max(offset_pct(last, baseline_avg));
        }
    }
    max_offset.max(0.0)
}

fn baseline_warning_status(tail_hit: bool, rise_hit: bool) -> String {
    match (tail_hit as u8) + (rise_hit as u8) {
        2 => "inaccurate".to_string(),
        1 => "warning".to_string(),
        _ => "ok".to_string(),
    }
}

// ── Effective baseline length ──────────────────────────────────────────────────

fn effective_baseline_points(
    sample_count: usize,
    requested_baseline_duration_sec: f64,
    apply_window_start_idx: usize,
) -> (usize, bool) {
    let requested = sample_count.min(seconds_to_points(requested_baseline_duration_sec.max(DT_SEC)).max(1));
    if requested <= apply_window_start_idx {
        return (requested, false);
    }
    let shortened = sample_count.min(apply_window_start_idx).max(1);
    (shortened, true)
}

// ── Drop detection ─────────────────────────────────────────────────────────────

fn first_threshold_crossing(samples: &[f64], threshold: f64, start: usize, end: usize) -> Option<usize> {
    if start >= end || end > samples.len() {
        return None;
    }
    samples[start..end].iter().position(|&v| v > threshold).map(|i| start + i)
}

fn fallback_drop_index(samples: &[f64], threshold: f64, baseline_end: usize) -> usize {
    let search_end = samples.len().min(baseline_end + LEGACY_DROP_SEARCH_POINTS);
    first_threshold_crossing(samples, threshold, baseline_end, search_end).unwrap_or(baseline_end)
}

fn build_timing_warning_details(
    requested_baseline_duration_sec: f64,
    effective_baseline_duration_sec: f64,
    baseline_was_auto_shortened: bool,
    apply_time_sec: f64,
    apply_tolerance_sec: f64,
    drop_search_fallback_used: bool,
) -> Vec<String> {
    let mut details = Vec::new();
    if baseline_was_auto_shortened {
        if effective_baseline_duration_sec <= DT_SEC {
            details.push(
                "Baseline overlapped the apply window and was clamped to the minimum baseline length."
                    .to_string(),
            );
        } else {
            details.push(format!(
                "Baseline shortened from {:.1}s to {:.1}s to stay before the apply window.",
                requested_baseline_duration_sec, effective_baseline_duration_sec
            ));
        }
    }
    if drop_search_fallback_used {
        details.push(format!(
            "No threshold crossing found within {:.1}s +/- {:.1}s; used automatic fallback search.",
            apply_time_sec, apply_tolerance_sec
        ));
    }
    details
}

// ── Public API ─────────────────────────────────────────────────────────────────

pub fn read_xlsx_single(path: &Path) -> Option<Vec<f64>> {
    let mut workbook = open_workbook_auto(path).ok()?;
    let range = workbook.worksheet_range_at(0)?.ok()?;
    let mut rows = range.rows();
    let header_row = rows.next()?;
    let column_index = header_row.iter().position(|cell| cell.to_string() == DATA_COL)?;
    let samples = rows.map(|row| row.get(column_index).map(cell_to_f64).unwrap_or(f64::NAN)).collect();
    Some(samples)
}

pub fn analyze_signal(data: &[f64], params: &AnalysisParams) -> Option<ProcessedSignal> {
    if data.is_empty() {
        return None;
    }
    let samples = data.to_vec();

    let requested_baseline_sec = params.baseline_duration_sec.max(DT_SEC);
    let apply_center_idx = seconds_to_points(params.drug_apply_time_sec.max(0.0));
    let apply_tolerance_pts = seconds_to_points(params.drug_apply_tolerance_sec.max(0.0));
    let apply_window_start_idx = apply_center_idx.saturating_sub(apply_tolerance_pts);
    let apply_window_end_idx = samples.len().min(apply_center_idx + apply_tolerance_pts + 1);

    let (baseline_len, baseline_was_auto_shortened) =
        effective_baseline_points(samples.len(), requested_baseline_sec, apply_window_start_idx);
    let baseline_segment = &samples[..baseline_len];
    let initial_avg = mean(baseline_segment);
    let std = sample_std(baseline_segment);
    let effective_baseline_duration_sec = baseline_len as f64 * DT_SEC;

    // Baseline accuracy
    let tail_len = baseline_segment.len().min(BASELINE_TAIL_POINTS);
    let tail_mean = if tail_len > 0 { mean(&baseline_segment[baseline_segment.len() - tail_len..]) } else { initial_avg };
    let tail_offset_pct = offset_pct(tail_mean, initial_avg);
    let rise_offset_pct = max_continuous_rise_offset_pct(baseline_segment, initial_avg);
    let warning_threshold = params.baseline_warning_threshold_pct.max(0.0);
    let tail_warning_hit = tail_offset_pct.abs() >= warning_threshold;
    let rise_warning_hit = rise_offset_pct >= warning_threshold;

    // Drop detection
    let threshold = initial_avg + DROP_SIGMA_THRESHOLD * std.max(1e-4);
    let window_search_start = baseline_len.max(apply_window_start_idx);
    let idx_drop_opt =
        first_threshold_crossing(&samples, threshold, window_search_start, apply_window_end_idx);
    let drop_search_fallback_used = idx_drop_opt.is_none();
    let (idx_drop, drop_detection_source) = if let Some(idx) = idx_drop_opt {
        (idx, "window".to_string())
    } else {
        (fallback_drop_index(&samples, threshold, baseline_len), "fallback_auto".to_string())
    };

    let final_avg = if samples.len() >= 100 { nanmean(&samples[samples.len() - 100..]) } else { f64::NAN };
    let delta_capacitance = if final_avg.is_nan() { f64::NAN } else { final_avg - initial_avg };

    let timing_warning_details = build_timing_warning_details(
        requested_baseline_sec,
        effective_baseline_duration_sec,
        baseline_was_auto_shortened,
        params.drug_apply_time_sec,
        params.drug_apply_tolerance_sec,
        drop_search_fallback_used,
    );

    Some(ProcessedSignal {
        time_sec: (0..samples.len()).map(|i| i as f64 * DT_SEC).collect(),
        capacitance: samples,
        drop_time: idx_drop as f64 * DT_SEC,
        delta_capacitance,
        initial_avg,
        effective_baseline_points: baseline_len,
        effective_baseline_duration_sec,
        baseline_was_auto_shortened,
        drop_detection_source,
        drop_search_fallback_used,
        timing_warning_details,
        baseline_warning_status: baseline_warning_status(tail_warning_hit, rise_warning_hit),
        baseline_tail_offset_pct: tail_offset_pct,
        baseline_rise_offset_pct: rise_offset_pct,
        baseline_tail_warning_hit: tail_warning_hit,
        baseline_rise_warning_hit: rise_warning_hit,
    })
}

pub fn process_single_file(path: &Path, params: &AnalysisParams) -> Option<ProcessedSignal> {
    let values = read_xlsx_single(path)?;
    if values.is_empty() {
        return None;
    }
    analyze_signal(&values, params)
}

// ── Tests ──────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use approx::assert_relative_eq;

    use super::analyze_signal;
    use crate::config::{DT_SEC, INITIAL_BASELINE_POINTS};
    use crate::models::AnalysisParams;

    fn default_params() -> AnalysisParams {
        AnalysisParams::default()
    }

    #[test]
    fn baseline_uses_configured_window_for_long_signal() {
        let mut samples = vec![10.0; INITIAL_BASELINE_POINTS];
        samples.extend(vec![13.0; 120]);
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_relative_eq!(signal.initial_avg, 10.0);
        assert_relative_eq!(signal.drop_time, INITIAL_BASELINE_POINTS as f64 * DT_SEC);
    }

    #[test]
    fn short_signal_uses_full_length_as_baseline() {
        let samples = vec![2.0, 4.0, 6.0, 8.0];
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_relative_eq!(signal.initial_avg, 5.0);
        assert!(signal.delta_capacitance.is_nan());
        assert_relative_eq!(signal.drop_time, samples.len() as f64 * DT_SEC);
    }

    #[test]
    fn threshold_crossing_uses_first_crossing_after_baseline() {
        let mut samples = vec![5.0; INITIAL_BASELINE_POINTS];
        samples.extend(vec![5.0, 5.0, 5.0, 8.5, 9.0]);
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_relative_eq!(signal.drop_time, (INITIAL_BASELINE_POINTS as f64 + 3.0) * DT_SEC);
    }

    #[test]
    fn no_crossing_falls_back_to_baseline_boundary() {
        let samples = vec![7.0; INITIAL_BASELINE_POINTS + 10];
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_relative_eq!(signal.drop_time, INITIAL_BASELINE_POINTS as f64 * DT_SEC);
    }

    #[test]
    fn delta_uses_last_hundred_points_when_available() {
        let mut samples = vec![10.0; INITIAL_BASELINE_POINTS];
        samples.extend(vec![11.0; 50]);
        samples.extend(vec![14.0; 100]);
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_relative_eq!(signal.delta_capacitance, 4.0);
    }

    #[test]
    fn drop_detection_source_is_window_when_crossing_in_apply_window() {
        let mut samples = vec![5.0; INITIAL_BASELINE_POINTS];
        samples.extend(vec![5.0, 9.0]); // crossing at index 201, within [200, 301]
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_eq!(signal.drop_detection_source, "window");
        assert!(!signal.drop_search_fallback_used);
    }

    #[test]
    fn drop_detection_source_is_fallback_when_no_window_crossing() {
        let samples = vec![7.0; INITIAL_BASELINE_POINTS + 10];
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert_eq!(signal.drop_detection_source, "fallback_auto");
        assert!(signal.drop_search_fallback_used);
    }

    #[test]
    fn baseline_tail_warning_fires_above_threshold() {
        // baseline rising over last 20 points → tail_offset_pct > 2%
        let mut samples: Vec<f64> = (0..INITIAL_BASELINE_POINTS)
            .map(|i| if i < 180 { 10.0 } else { 10.5 })
            .collect();
        samples.extend(vec![10.5; 100]);
        let signal = analyze_signal(&samples, &default_params()).unwrap();
        assert!(signal.baseline_tail_warning_hit);
    }
}
